### ISSUE TITLE
[ENH] Allow function pointers in Python code

### ISSUE THREAD

**Issue by scoder** [cython/cython#4279]
**Is your feature request related to a problem? Please describe.**
There is currently no way to define function pointers in pure Python syntax, e.g.
```cython
int (*grail)(int, char*) except -1
```

**Describe the solution you'd like**
open for debate

**Comment by da-woods** [comment]
I don't think we should use `typing.Callable` (because I think it implies a Python capable to me), but it might be consistent to borrow it's syntax?

**Comment by TeamSpen210** [comment]
For the `except` declarations (and any Cython-specific markers in general), perhaps that could be done by passing a call to the `exceptval()` decorator as a third "index"?
```py
grail = cython.FuncPointer[[int, pointer(char)], int, exceptval(-1)]
```
The mismatch between typing-style indexing and function calls does look rather odd though. Abusing slice syntax to do `"return": int` might also be reasonable, since the return value isn't very well identified here.

**Comment by scoder** [comment]
> the return value isn't very well identified here.

I'd rather rely on keyword arguments for the readability.

We can easily require some kind of strict spelling, even if Python syntax does not enforce it by itself, since users will want to run it through Cython in the end.

**Comment by erezinman** [comment]
@scoder Why not use the same syntax as `cython.typedef`?  For example:

`ctypedef int (*grail)(int, char*) except -1`

becomes:

`grail = cython.typedef(cython.int, function=True)(cython.int, cython.pointer(cython.char), _except=-1)`

This is relatively similar to the `ctypedef` statement, and easy to implement in Python (just add `__new__` to `Shadow.typedef` that returns a different class when the argument `function` is given). 

Another option is to lose the `function` argument, but than you might need to check the source code for the calling line of code (e.g. using the frame or [traceback](https://stackoverflow.com/a/72817601/2642356)) whether this is a function definition, or a regular type.

**Comment by dgrunwald-qt** [comment]
I'm currently attempting to implement something like this.

My suggestion:
```
cython.pointer[cython.function_type([cython.int], cython.double)]   # double (*)(int) except *
cython.pointer[cython.function_type([cython.int], cython.double, nogil=True)]  # double (*)(int) except * nogil
cython.pointer[cython.function_type([cython.int], cython.double, has_varargs=True)]  # double (*)(int, ...) except *
cython.pointer[cython.function_type([cython.int], cython.double, noexcept=True)]  # double (*)(int) noexcept
cython.pointer[cython.function_type([cython.int], cython.double, except_plus=True)]  # double (*)(int) except +
cython.pointer[cython.function_type([cython.int], cython.double, except_val=-1)]  # double (*)(int) except? -1
```

**Comment by scoder** [comment]
Time for bikeshedding. Alternatives for `except_plus`:
- `exceptval='+'` (It cannot be combined with other exception values anyway.)
- `except_cpp=True`

Since `except +` is not currently needed for _implementing_ functions, I don't think it's available yet in Python code, so this is a new piece of ground.

**Comment by scoder** [comment]
> `cython.pointer[cython.function_type([cython.int], cython.double, except_val=-1)]  # double (*)(int) except? -1`

This is not completely the same. `except_val=-1` should correspond to `except -1`. I propose to add `check_exception=True` for the `except?` case.

Note: The decorator is called `exceptval`, so I prefer to call the option likewise, not `except_val`.

**Comment by da-woods** [comment]
> `exceptval='+'` (It cannot be combined with other exception values anyway.)

You can at least combine it with `*`; I'm not sure about anything else.

https://github.com/cython/cython/blob/f8b2fcc1014bc9bdf2a849b997a1bd03fd6a8918/tests/run/cpp_exceptions.pyx#L25

**Comment by dgrunwald-qt** [comment]
`exceptval='+'` might be confused for `except '+'`. But if exception values won't ever be strings, we can use this.

In my current implementation, `cython.pointer[cython.function_type([cython.int], cython.double, noexcept=True, except_val=-1)]` will end up as `double (*)(int) except -1`; and without the `noexcept=True` you get `except? -1`.
But I guess we can accept either `check_exception` or `noexcept` (with an error if both are specified), and have `exceptval=-1` without either of them default to `except -1`. Then the new syntax would somewhat match the pyx syntax.

Putting it all together:
```
cython.pointer[cython.function_type([cython.int], cython.double)]   # double (*)(int) except *
cython.pointer[cython.function_type([cython.int], cython.double, nogil=True)]  # double (*)(int) except * nogil
cython.pointer[cython.function_type([cython.int], cython.double, has_varargs=True)]  # double (*)(int, ...) except *
cython.pointer[cython.function_type([cython.int], cython.double, noexcept=True)]  # double (*)(int) noexcept
cython.pointer[cython.function_type([cython.int], cython.double, exceptval='+')]  # double (*)(int) except +
cython.pointer[cython.function_type([cython.int], cython.double, exceptval='+', check_exception=True)]  # double (*)(int) except +*
cython.pointer[cython.function_type([cython.int], cython.double, exceptval=-1)]  # double (*)(int) except -1
cython.pointer[cython.function_type([cython.int], cython.double, exceptval=-1, check_exception=True)]  # double (*)(int) except? -1
cython.pointer[cython.function_type([cython.int], cython.double, exceptval=-1, noexcept=False)]  # double (*)(int) except? -1
```

**Comment by scoder** [comment]
> `exceptval='+'` might be confused for `except '+'`. But if exception values won't ever be strings, we can use this.

We could probably run into user code wanting to do this:
```cython
cython.pointer[cython.function_type([cython.int], cython.char, exceptval='+')]  # char (*)(int) except '+'
```
So, yeah, I don't think this is a safe API in the end. Let's have a dedicated spelling for C++ exception handling.

> In my current implementation, `cython.pointer[cython.function_type([cython.int], cython.double, noexcept=True, except_val=-1)]` will end up as `double (*)(int) except -1`; and without the `noexcept=True` you get `except? -1`.

`noexcept` is the explicit way to say that a function never raises an exception, so this doesn't fit here. `except?` really needs to be spelled out with an explicit exception check option as it is orthogonal to the exception value and both are the opposite of `noexcept`.

**Comment by dgrunwald-qt** [comment]
New version in updated PR:
```
cython.pointer[cython.function_type([cython.int], cython.double)]   # double (*)(int) except *
cython.pointer[cython.function_type([cython.int], cython.double, nogil=True)]  # double (*)(int) except * nogil
cython.pointer[cython.function_type([cython.int], cython.double, has_varargs=True)]  # double (*)(int, ...) except *
cython.pointer[cython.function_type([cython.int], cython.double, noexcept=True)]  # double (*)(int) noexcept
cython.pointer[cython.function_type([cython.int], cython.double, check_exception=False)]  # double (*)(int) noexcept
cython.pointer[cython.function_type([cython.int], cython.double, except_plus=True)]  # double (*)(int) except +
cython.pointer[cython.function_type([cython.int], cython.double, exceptval=-1)]  # double (*)(int) except -1
cython.pointer[cython.function_type([cython.int], cython.double, exceptval=-1, check_exception=True)]  # double (*)(int) except? -1
```

**Comment by scoder** [comment]
> `cython.pointer[cython.function_type([cython.int], cython.double, check_exception=False)]  # double (*)(int) noexcept`

This should map to `double (*)(int)`. I don't remember right now if we defaulted this to `except -1.0` or an implicit `noexcept`. In any case, the `check_exception=False` should have no effect in this case, it's redundant, because a `double` return has a known error return value and doesn't need an additional exception check.

**Comment by scoder** [comment]
> I don't remember right now if we defaulted this to `except -1.0` or an implicit `noexcept`

I looked it up in the ticket: https://github.com/cython/cython/issues/4280#issuecomment-1100592792
We decided at the time that function pointers should behave like Cython implemented function signatures, i.e. use the default exception value of `except? -1.0` for a `double` return value, and with the additional `check_exception=False` option, turn into `except -1.0`.

(This does not apply to function pointers in `cdef extern` blocks, but these cannot currently appear in Python code anyway.)

**Comment by da-woods** [comment]
I'd probably go for `except_cpp` rather than `except_plus`. In that the intent is clear without knowing the traditional Cython syntax (where I guess `except +` is mainly an attempt to do something without adding too much new syntax).

>> `cython.pointer[cython.function_type([cython.int], cython.double, check_exception=False)]  # double (*)(int) noexcept`
>
> We decided at the time that function pointers should behave like Cython implemented function signatures, i.e. use the default exception value of `except? -1.0` for a double return value, and with the additional `check_exception=False` option, turn into except -1.0.

A quick test tells me that

```
ctypedef double (*some_ptr_type)()
```

is currently interpreted as `except *`

To me that seems consistent with

```
cython.pointer[cython.function_type([cython.int], cython.double)]   # double (*)(int) except *
cython.pointer[cython.function_type([cython.int], cython.double, check_exception=False)]  # double (*)(int) noexcept
```

I'd prefer keeping the behaviour as consistent as possible with the `cdef` syntax, so the list that @dgrunwald-qt proposes looks right to me.

**Comment by scoder** [comment]
> `except_cpp` rather than `except_plus`.

+1

> keeping the behaviour as consistent as possible with the `cdef` syntax

Absolutely. I think a direct comparison test would help assuring that and keeping it that way.

**Comment by clayote** [comment]
The way I naïvely tried to annotate this was `cython.pointer[cython.cfunc[[cython.int], cython.void]]`. I don't know what problems might be caused by overloading the `cfunc` decorator like that, but I suppose it does mean you can do the same to `ccall`.
